Compute the entropy of the English alphabet assuming uniform distribution (26 letters). Then estimate it using actual letter frequencies. Which is higher and why?

In [1]:
import numpy as np

# Uniform distribution: 26 letters
p_uniform = np.ones(26) / 26

entropy_uniform = -np.sum(
    p_uniform * np.log2(p_uniform)
)

print(f"Uniform entropy: {entropy_uniform:.4f} bits")

Uniform entropy: 4.7004 bits


In [3]:
import numpy as np

letters = np.array([
    8.17, 1.49, 2.78, 4.25, 12.70, 2.23, 2.02,
    6.09, 6.97, 0.15, 0.77, 4.03, 2.41, 6.75,
    7.51, 1.93, 0.10, 5.99, 6.33, 9.06, 2.76,
    0.98, 2.36, 0.15, 1.97, 0.07
])

p_actual = letters / 100

entropy_actual = -np.sum(
    p_actual * np.log2(p_actual)
)

print(f"Actual English entropy: {entropy_actual:.4f} bits")

Actual English entropy: 4.1762 bits


**Why?**

- Entropy measures uncertainty.

- With the uniform distribution:

`
A → 1/26
B → 1/26
C → 1/26
...
Z → 1/26
`

- Every letter is equally likely, so you're maximally uncertain about which letter comes next.

- But in English:
`
E → ~12.7%
T → ~9.1%
A → ~8.2%
...
`
- Some letters are much more common than others.

So if I ask:

> "What letter comes next?"

knowing that it's English gives you some information because **E, T, A, O**, etc. are more likely.

Therefore uncertainty decreases:

$$ \boxed{4.70 > 4.18} $$

---

### Tiny memory trick

> **Uniform = maximum uncertainty = maximum entropy.**
> **Uneven probabilities = more predictable = lower entropy.**

Show that KL divergence is not symmetric. Pick two distributions P and Q and compute D_KL(P || Q) and D_KL(Q || P). Explain why they differ.

In [4]:
import numpy as np

def kl_divergence(P, Q):
    P = np.asarray(P, dtype=float)
    Q = np.asarray(Q, dtype=float)

    return np.sum(P * np.log2(P / Q))


P = np.array([0.8, 0.2])
Q = np.array([0.5, 0.5])

kl_pq = kl_divergence(P, Q)
kl_qp = kl_divergence(Q, P)

print(f"D_KL(P || Q) = {kl_pq:.4f} bits")
print(f"D_KL(Q || P) = {kl_qp:.4f} bits")

D_KL(P || Q) = 0.2781 bits
D_KL(Q || P) = 0.3219 bits
